### PR TITLE
Add a low memory limit hint

### PR DESCRIPTION
**Pull request by dcodeIO**
This is an initial implementation of a `--lowMemoryLimit` hint as suggested in https://github.com/AssemblyScript/assemblyscript/issues/1089.

* Emits an error if the limit is exceeded by static data
* <del>Emits an error</del> Traps if the limit is exceeded by TLSF bookkeeping structures
* Restricts initial memory added to TLSF's free list
* Only useful if available memory is less than 64kb

This might have several corner cases, but should be good for initial test runs to find out more.

### CODE REVIEW

File: std/assembly/rt/tlsf.ts [hunk 1]
```diff
@@ -475,7 +475,14 @@ export function maybeInitialize(): Root {
...
     }
-    addMemory(root, (rootOffset + ROOT_SIZE + AL_MASK) & ~AL_MASK, memory.size() << 16);
+    if (ASC_LOW_MEMORY_LIMIT > 0) {
+      const start = (rootOffset + ROOT_SIZE + AL_MASK) & ~AL_MASK;
+      const end = (<usize>ASC_LOW_MEMORY_LIMIT + AL_MASK) & ~AL_MASK;
+      if (start > end) ERROR("Low memory limit exceeded by TLSF bookkeeping");
+      else if (start < end) addMemory(root, start, end);
+    } else {
+      addMemory(root, (rootOffset + ROOT_SIZE + AL_MASK) & ~AL_MASK, memory.size() << 16);
+    }
```
Comment (MaxGraey) [inline std/assembly/rt/tlsf.ts]: Simplify?
```ts
    const start = (rootOffset + ROOT_SIZE + AL_MASK) & ~AL_MASK;
    let end: usize;
    if (ASC_LOW_MEMORY_LIMIT > 0) {
      end = (<usize>ASC_LOW_MEMORY_LIMIT + AL_MASK) & ~AL_MASK;
      if (start > end) ERROR("Low memory limit exceeded by TLSF bookkeeping");
    } else {
      end = memory.size() << 16;
    }
    addMemory(root, start, end);
```

Comment (dcodeIO) [reply std/assembly/rt/tlsf.ts]: Looks like `let end: usize;` would break precompute there.

Comment (MaxGraey) [reply std/assembly/rt/tlsf.ts]: Ah, right. I forgot what it is conditional compilation)

Comment (dcodeIO) [reply std/assembly/rt/tlsf.ts]: But now that you mentioned it, it seems `start` does not precompute anyway due to depending on `__heap_base`. Hmm...

**Comment by dcodeIO** [comment]
Iirc this didn't work perfectly the last time it was tried, but also isn't breaking anything. As such I suggest to merge so more people with a respective use case can get their hands on it, and improve once we know more.
